New topic Insights & analytics (#15373)
docker/docker.github.io: docker-hub/publish/insights-analytics.md

---

> title: Insights & analytics

Reviewer: Do we want to add anything to the DVP overview page that references/links to this page?

Reviewer 2: I will update the DVP overview page in a separate PR, so I will make sure to include a link to this page then

and get breakouts by geolocation, cloud provider, and tool. The data is
available as a downloadable CSV file, in a weekly (Monday through Sunday) or
monthly format (available on the first day of the following calendar month). You
can import this data into your own systems or you can analyze it manually as a
spreadsheet. Review the [Data definitions](#data-definitions) section for more
information about how to read the data contained in the CSV file.

There are automated systems that frequently check for new versions of your
images. The Insights and Analytics metrics show the number of pulls that were
triggered by users, and automated machines such as CI/CD systems, respectively.
We have tried to distinguish between "version checks" and real image downloads
by inspecting the order and timing of image pulls coming from the same IP
address. Being able to distinguish between different types of image pulls grants
you more insight into your user's behavior. You can inspect the rules we have
applied for determining intent behind pulls in the
[Action classification rules](#action-classification-rules) section on this
page. We welcome [feedback](https://forms.gle/nb7beTUQz9wzXy1b6){:

> target="_blank" rel="noopener" class="_"} and questions on these rules.

Reviewer 2: This paragraph still needs some love.

## Export analytics data from the web

Here's how to export usage data for your organization's images using the Docker
Hub website.

> 1. Log in to [Docker Hub](https://hub.docker.com/){: target="_blank"

Author: Log in to Docker Hub and click Organizations.

   rel="noopener" class="_"}.

2. Click **Organizations** in the top navigation bar, then choose your
   organization from the list.

> 3. On your organization's landing page, click the **Insights and analytics**

Author: Choose your organization and click Insights and analytics.

2. Choose your organization and click **Insights and analytics**.

   ![Organization overview page, with the Insights and Analytics tab](./images/organization-tabs.png)

3. Set the time span for which you want to export analytics data. The
   downloadable CSV files for summary and raw data appear on the right-hand
   side.

   ![Filtering options and download links for analytics data](./images/download-analytics-data.png)

> ## Raw data

Reviewer: I wonder if it's worth looking at the heading structure. It feels a little bit scattergun currently. Maybe we could group everything under 3 main sections - What is the tool?, How do I use the tool? and What data can I get? (obviously with way better names)

Reviewer 2: yep for sure, it's all over the place. Let me take a stab at it.

Reviewer: 👍

Reviewer 2: @chrispatrick do you reckon it's any better now?

Reviewer: Yeah, I think that works better

---
title: Insights & analytics
description: Provides usage statistics of your images on Docker Hub.
keywords: docker hub, hub, insights, analytics, api, verified publisher
---

Insights & analytics provides usage analytics for your organization's images on
Docker Hub. With this tool, you have self-serve access to metrics as both raw
data and summary data for a desired time span. You can view how many times your

> images have been pulled by tag or by digest, and get breakouts by geolocation,

Reviewer 2: you're right, in this case it's probably breakdowns!

Sunday) or monthly format (available on the first day of the following calendar
month). You can import this data into your own systems, or you can analyze it
manually as a spreadsheet.

### Export data using the website

Here's how to export usage data for your organization's images using the Docker
Hub website.

1. Log in to [Docker Hub](https://hub.docker.com/){: target="_blank"
   rel="noopener" class="_"} and select **Organizations**.

2. Choose your organization and click **Insights and analytics**.

   ![Organization overview page, with the Insights and Analytics tab](./images/organization-tabs.png)

3. Set the time span for which you want to export analytics data. The
   downloadable CSV files for summary and raw data appear on the right-hand
   side.

   ![Filtering options and download links for analytics data](./images/download-analytics-data.png)

### Export data using the API

The HTTP API endpoints are available at:
`https://hub.docker.com/api/publisher/analytics/v1`. Learn how to export data
using the API in the [DVP Data API documentation](/docker-hub/api/dvp/){:
target="_blank" rel="noopener" class="_"}.

> ```bash

Reviewer 2: yep... nothing 😄

## Action classification rules

Automated systems frequently check for new versions of your images. The insights
and analytics metrics show the number of pulls that were triggered by users, and
pulls by automated systems such as CI/CD tools, respectively. Automated "version
checks" and real image downloads are differentiated by inspecting the order and
timing of image pulls coming from the same IP address. Being able to distinguish
between different types of image pulls grants you more insight into your users'
behavior. You can inspect the rules for determining intent behind pulls in the
[Action classification rules](#action-classification-rules) section on this
page.

> Please let us know if you have any

Reviewer 3: ```suggestion
To provide feedback or ask questions about these rules, [fill out the Google Form]((https://forms.gle/nb7beTUQz9wzXy1b6){: target="_blank" rel="noopener"
class="_"}
```

---
title: Insights & analytics

> description: Provides an API analytics tool for your products.

Reviewer: We should probably reference the UI that can be used as well as the raw API.

Reviewer 4: Agreed!

Reviewer 4: We should also add screenshots of the UI in this documentation

Reviewer 2: OH yeah. I was totally focused on the API, probably because we merged the API doc updates! Let me add some information about the GUI view.

Reviewer 2: Updated with info about the UI -- please take a look!

keywords: docker hub, hub, insights, analytics, api
---

> The Insights & Analytics tool includes an API that provides analytics for your products. With this API, you now have self-serve access to metrics where you can download both raw data and summary data. You can view how many times your images have been pulled by tag or by digest, and get breakouts by geolocation, cloud provider, and tool.The API measures the number of pulls that were triggered by users, and automated machines such as CI/CD systems. You can import this weekly or monthly data into your own systems or you can analyze it manually using spreadsheet reports. The data can then be exported as a downloadable GZipped CSV file. Review the Data Definitions section for more information.

Reviewer: ```suggestion
The Insights & Analytics tool includes an API that provides analytics for your products. With this API, you now have self-serve access to metrics where you can download both raw data and summary data. You can view how many times your images have been pulled by tag or by digest, and get breakouts by geolocation, cloud provider, and tool. The API measures the number of pulls that were triggered by users, and automated machines such as CI/CD systems. You can import this weekly or monthly data into your own systems or you can analyze it manually using spreadsheet reports. The data can then be exported as a downloadable GZipped CSV file. Review the Data Definitions section for more information.
```

There are automated systems that frequently check for new versions of your images, and we have tried to distinguish between those 'version checks' and real user downloads by inspecting the order and timing of API calls from the same IP address. This will provide more insight into customer behavior. You can inspect the rules we have applied in the Action Classification Rule Set section of this document. We welcome [feedback](https://forms.gle/nb7beTUQz9wzXy1b6){: target="_blank" rel="noopener" class="_"} and questions on these rules.

## Raw data

> In your raw data CSV you will have access to the data points listed below. You can request raw data for a complete week (Monday through Sunday) or for a complete month (available on the first day of the following month). **Note:** each action is represented as a single row.

Reviewer 4: In your raw data CSV you will have access to the data points listed below. You can request raw data for a complete week (Monday through Sunday) or for a complete month (available on the first day of the following month). **Note:** each action is represented as a single row in the CSV file.

Reviewer 2: fixed ✅

  * Type (industry)
  * Host (cloud provider)
  * Country (geolocation)
  * Timestamp
  * Namespace
  * Repo
  * Reference (digest is always included, tag is provided when available)
  * HTTP method
  * Action, one of the following:
    * Pull by tag
    * Pull by digest
    * Version check
  * User-Agent Tool

## Summary data

In your summary data CSV, you will have access to the data points listed below. You can request summary data for a complete week (Monday through Sunday) or for a complete month (available on the first day of the following month).

For every namespace, repository, and reference (tag or digest):

  * Unique IP address count
  * Pulls by tag count
  * Pulls by digest count
  * Version check count

> 

Reviewer 4: We should add a section here about accessing the data via the UI. 

- where to access the UI (example: https://hub.docker.com/orgs/docker/insights)
- who has access to the UI and API (and team member of an organization)
- add screenshot of the UI

Reviewer 2: Hey Britney. I added a new section for the UI. Please check it out! And for who has access, I'm not sure if I got it right but I added this to the first paragraph:

All members of an organization has access to the analytics data.

## Data Definitions

> | Data Point | Definition |

Reviewer 2: ```suggestion
| Data point | Definition |
```

|:-----|:--------|

> | Action | An Action represents the multiple request events associated with a docker pull. We have applied rules to these events so that the data is more meaningful in analyzing user behavior and intent. An Action can be filtered into three distinct categories: version check, pull by tag, and pull by digest. Each Action is represented as a single row in the raw data. |

Reviewer 2: ```suggestion
| Action | An Action represents the multiple request events associated with a `docker pull`. We have applied rules to these events so that the data is more meaningful in analyzing user behavior and intent. An Action can be filtered into three distinct categories: version check, pull by tag, and pull by digest. Each Action is represented as a single row in the raw data. |
```

| Version check  |  This is a filter on the Action data point. It is a speculation of user intent. Includes: HEAD by tag not followed by a GET (from the same IP address within a 5-second window). Excludes: HEAD by digest |

> | Pull by tag | This is a filter on the Action data point. It is a speculation of user intent. Includes: GET (by digest or by tag). If the GET is immediately preceded by a HEAD by tag (from the same IP address within a 5-second window), then the GET and HEAD together are counted as a single Pull by Tag. If the GET by tag is immediately followed by another GET (from the same IP address within a 5-second window, but a different digest), then the two GETs are counted as a single Pull by Tag. |

Reviewer: The data here feels a little duplicated with the `Action classification ruleset` below... Not a big deal though.

|Pull by Digest  |This is a filter on the Action data point. It is a speculation of user intent. Includes: GET by digest. If the GET is immediately preceded by a HEAD by digest (by the same IP address within a 5-second window), then the GET and HEAD together are counted as 1 pull by digest. If the GET is immediately followed by another GET (same IP address within a 5-second window, but a different digest), then both GETs together are counted as 1 pull by digest. Includes: HEAD by digest, not followed by a GET|
|Type | The industry from which the event originates. Industry types include “business”, “ISP” (internet service provider), “hosting”, “education”, and “null” in cases where the industry could not be identified.|
|Host |The cloud service provider used in an event.|
|Country |The country from which the request originated.  |
|Timestamp |Date & time of an event in the following schema: YYYY-MM-DD 00:00:00|
|Namespace |The Docker organization that a repository is a part of. |
|Repo |The repository that an image belongs to.|
|Reference |The tag or digest of any given image.|

> |HTTP Method | The HTTP method(s) used in a request by the client. More information on Docker Registry HTTP API protocols can be found here.|

Reviewer 2: ```suggestion
|HTTP Method | The HTTP method used in a request by the client. More information on Docker Registry HTTP API protocols can be found [here](/registry/spec/api/){: target="_blank" rel="noopener" class="_"}.|
```

> |User-Agent Tool|The application a user used to pull an image.  Extracted from the UA string.|

Reviewer 2: I added User Agent Version because I can see that in the CSV I export from the GUI. I can also see "Action Day", but I am not seeing the "Unique IP Addresses" counter in the CSV. Maybe we need to double check that these data points are correct?

```suggestion
|User Agent Tool|The application a user used to pull an image.  Extracted from the UA string.|
|User Agent Version|The application version, extracted from the UA string.|
```

Reviewer: The columns vary a little between raw data and summary data. For instance "Unique IP Addresses" only makes sense in the summary data, since the raw data is made up of single actions originating from single IPs.

Probably worth adding `Action Day` in though. I'm not sure it's the best name, but it just refers to the date portion of the timestamp.

|Unique IP Address|As part of our privacy-preserving policy, Docker only shares the count of distinct unique IP addresses that request an image.|

> ## Action Classification Rule Set

Reviewer 2: ```suggestion
## Action classification ruleset
```

> |Starting Event | Reference | Followed By | Resulting Action | Use Case(s) |Notes|

Reviewer 2: ```suggestion
|Starting event | Reference | Followed by | Resulting action | Use case(s) |Notes|
```

|:-----|:--------|:-----|:--------|:-----|:--------|

> | HEAD | tag | N/A | Version Check | User already has all layers existing on local machine|This is similar to the use case of a pull by tag when the user already has all the image layers existing locally, however, we are able to differentiate the user intent and classify accordingly.|

Reviewer 2: ```suggestion
| HEAD | tag | N/A | Version check | User already has all layers existing on local machine|This is similar to the use case of a pull by tag when the user already has all the image layers existing locally, however, we are able to differentiate the user intent and classify accordingly.|
```

> | Get| tag | N/A |Pull by Tag | User already has all layers existing on local machine and / or Image is single arch |

Reviewer 2: ```suggestion
| GET| tag | N/A |Pull by tag | User already has all layers existing on local machine and / or Image is single arch |
```

> | Get | tag | Get by different digest | Pull by tag | Image is multiarch |Following GET by digests must be different from the first digest|

Reviewer 2: ```suggestion
| GET | tag | Get by different digest | Pull by tag | Image is multi-arch |Second GET by digests must be different from the first|
```

> | HEAD | tag | GET by same digest | Pull by Tag | Image is multi-arch but some part of the image already exists on the local machine and/or User already has all layers existing on local machine|The HEAD by tag will send the most current digest, the following GET must be by that same digest. There may be an additional GET, if the image is multi arch (use case on the next line). If the user doesn't want the most current digest, then the user would perform HEAD by digest.|

Reviewer 2: ```suggestion
| HEAD | tag | GET by same digest | Pull by tag | Image is multi-arch but some or all image layers already exist on the local machine. |The HEAD by tag will send the most current digest, the following GET must be by that same digest. There may occur an additional GET, if the image is multi-arch (see the next row in this table). If the user doesn't want the most recent digest, then the user would perform HEAD by digest.|
```

> | HEAD| tag |GET by same digest, THEN a second GET by different digest| Pull by Tag| Image is multi-arch |The HEAD by tag will send the most current digest, the following GET must be by that same digest. Since the image is multi arch, there is a second GET by a different digest. If the user doesn't want the most current digest, then the user would perform HEAD by digest.|

Reviewer 2: ```suggestion
| HEAD| tag |GET by the same digest, then a second GET by a different digest| Pull by tag| Image is multi-arch |The HEAD by tag will send the most recent digest, the following GET must be by that same digest. Since the image is multi-arch, there is a second GET by a different digest. If the user doesn't want the most recent digest, then the user would perform HEAD by digest.|
```

> | HEAD | tag | GET by same digest, THEN a second GET by different digest | Pull by Tag| Image is multi arch | The HEAD by tag will send the most current digest, the following GET must be by that same digest. Since the image is multi-arch, there is a second GET by a different digest. If the user doesn't want the most current digest, then the user would perform HEAD by digest.|

Reviewer 2: ```suggestion
| HEAD | tag | GET by same digest, then a second GET by different digest | Pull by tag| Image is multi-arch | The HEAD by tag will send the most current digest, the following GET must be by that same digest. Since the image is multi-arch, there is a second GET by a different digest. If the user doesn't want the most recent digest, then the user would perform HEAD by digest.|
```

| GET | digest | N/A | Pull by Digest |  User already has all layers existing on local machine and/or Image is single arch |

> | HEAD | digest | N/A | Pull by Digest | User already has all layers existing on local machine |

Reviewer 2: ```suggestion
| HEAD | digest | N/A | Pull by digest | User already has all layers existing on their local machine. |
```

> | GET | digest | GET by different digest | Pull by Digest | Image is multi arch |The following GET by digest must be different from the first digest|

Reviewer 2: ```suggestion
| GET | digest | GET by different digest | Pull by Digest | Image is multi-arch |The second GET by digest must be different from the first|
```

| HEAD | digest | GET by same digest | Pull by digest |  Image is single arch and/or image is multi-arch but some part of the image already exists on the local machine|

> | Head | digest | GET by same digest, tHEN a second GET by different digest | Pull by Digest | Image is multi arch |

Reviewer 2: ```suggestion
| HEAD | digest | GET by same digest, then a second GET by different digest | Pull by Digest | Image is multi-arch |
```
